### ISSUE TITLE
lightning writeup

### ISSUE THREAD

**Issue by prakashjayy** [prakashjayy/fullstack_dl#7]
- To debug use `import pdb; pdb.set_trace()`

```python
import pytorch_lightning as pl
```

### Main components 
1. Model
2. optimizer 
3. data
4. training loop "the magic"
5. validation loop "the validation magic"

```python
class ResNet(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.l1 = nn.Linear()
        self.loss = nn.CrossEntropyLoss()

    def forward(self, x):
        return logits 

    def configure_optimizers():
        return optimizer 

    def training_step(self, batch, batch_idx):
        acc = accuracy(logits, y)
        pbar = {"train_acc": acc}
        return {"loss": loss, "progress_bar": pbar} 

    def validation_step(self, batch, batch_idx):
        avg_val_loss =  self.training_step(batch, batch_idx)
        return {"val_loss": avg_val_loss} 

    def prepare_data(self):
        return 

    def setup(self):
       self.train, self.val = ...

   def validation_epoch_end(self, val_step_outputs):
       return torch.tensor([x["loss"] for x in val_step_outputs]).mean()

    def train_dataloader(self):
       return train_loader

    def val_dataloader(self):
       return val_dataloader

model = ResNet()
trainer = pl.Trainer()
trainer.fit(model)
```

- `progress_bar`, "val_loss",  and `loss` are reserved words in pl
- don't download data in train_dataloader as multi-gpu setup will download it multiple times, so use `prepare_data` and `setup` methods.
- decentralize datamodule so that lightning trainer need not worry about what is there in the data. 🆗 
- when defining a dataloader, set `num_workers` equal to the number of CPUs on that machine and `pin_memory` to True.
- `seed_everthing(42)` and set `deterministic=True` for **reproducibility** .  💁‍♂️ 
- `auto_lr_find=True`  on Trainer and then do `trainer.tune(model)` for checking cyclical learning rate
- set `benchmark=True` will increase the speed if your batch_size is same across all iteration. 
- [callbacks](https://pytorch-lightning.readthedocs.io/en/latest/generated/pytorch_lightning.callbacks.Callback.html) to check for - [`ModelCheckpoint`](https://pytorch-lightning.readthedocs.io/en/latest/generated/pytorch_lightning.callbacks.ModelCheckpoint.html), [`EarlyStopping`](https://pytorch-lightning.readthedocs.io/en/latest/generated/pytorch_lightning.callbacks.EarlyStopping.html), `LearningRateMonitor`

## TODO
- amp_backend `native` or `apex` and its associated params like `amp_level` 🅰️
